Partitioned write into s3 table function

excitoon opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Partitioned write into s3 table function

Detailed description / Documentation draft:

...

Shall resolve #15171

excitoon committed:
Proper parser.

excitoon committed:
Parsing.

excitoon committed:
Partitioned writes for S3.

excitoon committed:
Test.

excitoon committed:
Fixes.

excitoon committed:
Better functionality.

excitoon commented:
Failed tests are unrelated.

vdimir commented on src/Storages/StorageS3.cpp:
```diff
@@ -320,6 +335,141 @@ class StorageS3BlockOutputStream : public IBlockOutputStream
+    }
+
+    void writePrefix() override
+    {
```
```suggestion
    {
        /// writeSuffix for each child writer called in getWriterForPartition on writer creation
```

vdimir commented on src/Storages/StorageS3.cpp:
```diff
@@ -320,6 +335,141 @@ class StorageS3BlockOutputStream : public IBlockOutputStream
+        Block current_block_with_partition_by_expr = block;
+        partition_by_expr->execute(current_block_with_partition_by_expr);
+
+        const auto * key_column = checkAndGetColumn<ColumnString>(current_block_with_partition_by_expr.getByName(partition_by_column_name).column.get());
```
Maybe we need to sanitize partition key in some way?

For example following code inserting string containing zero bytes

<details>
<summary>Python code snippet</summary>

```
from clickhouse_driver import Client
ch_client = Client('localhost', 9000)

command = """
INSERT INTO TABLE FUNCTION 
s3('http://localhost:9001/foo/test_{_partition_id}.csv', 
    'minioadmin', 'minioadmin', 'CSV', 'id Int32, val String') 
PARTITION BY val
VALUES
"""

ch_client.execute(command, [
    {'id': 1, 'val': '\x00\x00\x00\x00'},
])
```

</details>

Raises

<details>
<summary>ServerException</summary>

```
ServerException: Code: 499.
DB::Exception: The request signature we calculated does not match the signature you provided. Check your key and signing method.. Stack trace:

0. /build/build_docker/../contrib/libcxx/include/exception:133: Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x11fada0c in /home/vdimir/_tmp/clickhouse-23051
1. /build/build_docker/../src/Common/Exception.cpp:59: DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0x8f2ac9a in /home/vdimir/_tmp/clickhouse-23051
2. /build/build_docker/../src/IO/WriteBufferFromS3.cpp:233: DB::WriteBufferFromS3::makeSinglepartUpload() @ 0xf03d7e8 in /home/vdimir/_tmp/clickhouse-23051
3. /build/build_docker/../src/IO/WriteBufferFromS3.cpp:112: DB::WriteBufferFromS3::finalizeImpl() @ 0xf03ce0b in /home/vdimir/_tmp/clickhouse-23051
4. /build/build_docker/../src/IO/WriteBufferFromS3.cpp:0: DB::WriteBufferFromS3::finalize() @ 0xf03cd61 in /home/vdimir/_tmp/clickhouse-23051
5. /build/build_docker/../src/Storages/StorageS3.cpp:329: DB::StorageS3BlockOutputStream::writeSuffix() @ 0xf944b28 in /home/vdimir/_tmp/clickhouse-23051
6. /build/build_docker/../src/Storages/StorageS3.cpp:0: DB::StorageS3PartitionedBlockOutputStream::writeSuffix() @ 0xf942c3a in /home/vdimir/_tmp/clickhouse-23051
7. /build/build_docker/../contrib/libcxx/include/exception:176: DB::PushingToViewsBlockOutputStream::writeSuffix() @ 0xf357143 in /home/vdimir/_tmp/clickhouse-23051
8. /build/build_docker/../src/Server/TCPHandler.cpp:581: DB::TCPHandler::processInsertQuery(DB::Settings const&) @ 0xfccf69f in /home/vdimir/_tmp/clickhouse-23051
9. /build/build_docker/../contrib/libcxx/include/functional:2236: DB::TCPHandler::runImpl() @ 0xfcca3cd in /home/vdimir/_tmp/clickhouse-23051
10. /build/build_docker/../src/Server/TCPHandler.cpp:1623: DB::TCPHandler::run() @ 0xfcd5c19 in /home/vdimir/_tmp/clickhouse-23051
11. /build/build_docker/../contrib/poco/Net/src/TCPServerConnection.cpp:57: Poco::Net::TCPServerConnection::start() @ 0x11f215a7 in /home/vdimir/_tmp/clickhouse-23051
12. /build/build_docker/../contrib/poco/Net/src/TCPServerDispatcher.cpp:114: Poco::Net::TCPServerDispatcher::run() @ 0x11f21a3c in /home/vdimir/_tmp/clickhouse-23051
13. /build/build_docker/../contrib/poco/Foundation/include/Poco/ScopedLock.h:36: Poco::PooledThread::run() @ 0x120088d9 in /home/vdimir/_tmp/clickhouse-23051
14. /build/build_docker/../contrib/poco/Foundation/include/Poco/SharedPtr.h:156: Poco::ThreadImpl::runnableEntry(void*) @ 0x120065e6 in /home/vdimir/_tmp/clickhouse-23051
15. start_thread @ 0x76db in /lib/x86_64-linux-gnu/libpthread-2.27.so
16. /build/glibc-OTsEL5/glibc-2.27/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:97: __clone @ 0x12188f in /usr/lib/debug/lib/x86_64-linux-gnu/libc-2.27.so
```
</details>

It's not a big issue, but I'm not sure that something more dangerous can't happen. Any ideas?

excitoon replied:
There is no strong restrictions on key names in S3 :( And no code for validation in SDK. As for `\x00` symbol, it is not technically forbidden, probably this can be even fixed.

https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html

vdimir replied:
It's stated that `You can use any UTF-8 character in an object key name`, but we can generate invalid UTF-8 string. Anyway, there's `Safe characters`, do you mind to add validation according to this, throwing exception if not valid string found?

excitoon replied:
We already have people who use unsafe characters like `+` (specifically fixed that) and also there may be other non AWS S3 implementations🤷🏼‍♂️🤷🏼‍♂️

vdimir replied:
But my example above return error and message is not clear what happened

vdimir reviewed (commented):
Thank you for you changes!

I took a first look and left comments.

Also maybe I'll dig into details once again.

excitoon committed:
Review fixes.

vdimir commented on src/Storages/StorageS3.cpp:
```diff
@@ -42,6 +52,11 @@
+#include <boost/algorithm/string.hpp>
+
+
+static const String PARTITION_ID_WILDCARD = "{_partition_id}";
```
If string already "escaped" with `{}`, do we nedd `_` at the beginning. Looks like overkill, maybe just `{partition_id}`. But actually I'm not sure which is better.

excitoon replied:
I did that to match existing virtual column name, additionally saving
future possibility to implement expressions or formats (may be).

excitoon commented:
I did that to match existing virtual column name, additionally saving
future possibility to implement expressions or formats (may be).

On Wed, Jul 14, 2021, 3:47 PM Vladimir ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/Storages/StorageS3.cpp
> <https://github.com/ClickHouse/ClickHouse/pull/23051#discussion_r669582815>
> :
>
> > @@ -42,6 +52,11 @@
>
>  namespace fs = std::filesystem;
>
> +#include <boost/algorithm/string.hpp>
> +
> +
> +static const String PARTITION_ID_WILDCARD = "{_partition_id}";
>
> If string already "escaped" with {}, do we need _ at the beginning. Looks
> like overkill, maybe just {partition_id}. But actually I'm not sure which
> is better.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ClickHouse/ClickHouse/pull/23051#pullrequestreview-706230842>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARGSB26GIUHW5NZXDUFOBTTXWBP3ANCNFSM424C2WJA>
> .
>

vdimir committed:
Add test for unsupported INSERT PARTITION BY

vdimir commented on src/Parsers/ParserInsertQuery.cpp:
```diff
@@ -60,6 +64,12 @@ bool ParserInsertQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
+
+        if (s_partition_by.ignore(pos, expected))
+        {
+            if (!exp_elem_p.parse(pos, partition_by_expr, expected))
```
Maybe let's add test for it? I want to push it to PR, but seems that there's no permission. https://github.com/vdimir/ClickHouse/commit/70e87b3f46ccba6193345f707785e72bd285b6e7

excitoon replied:
Added commit and permission.

vdimir commented on src/Common/ErrorCodes.cpp:
```diff
@@ -561,6 +561,7 @@
     M(591, SQLITE_ENGINE_ERROR) \
     M(592, DATA_ENCRYPTION_ERROR) \
     M(593, ZERO_COPY_REPLICATION_ERROR) \
+    M(594, PARTITION_BY_NOT_SUPPORTED) \
```
```suggestion
    M(594, NOT_IMPLEMENTED) \
```

vdimir replied:
And I also cannot commit this via web interface. You can apply three suggestion or cherry pick https://github.com/ClickHouse/ClickHouse/commit/9c1cb601e957f96e5bbceb811bb0070060a63d5a

excitoon replied:
Pushed existing `NOT_IMPLEMENTED`.

vdimir commented on src/Interpreters/InterpreterInsertQuery.cpp:
```diff
@@ -40,6 +40,7 @@ namespace ErrorCodes
     extern const int NO_SUCH_COLUMN_IN_TABLE;
     extern const int ILLEGAL_COLUMN;
     extern const int DUPLICATE_COLUMN;
+    extern const int PARTITION_BY_NOT_SUPPORTED;
```
```suggestion
    extern const int NOT_IMPLEMENTED;
```

vdimir reviewed (commented):
Ok, generally it's good, except non-validated keys and unclear error message around it.

excitoon committed:
Review fixes.

vdimir committed:
Add validation to insert partition by key to s3

vdimir commented:
I've added partition key validation. We can change it later if it'll be too strict, I suppose it's better than adding restriction and broke some behavior in future.

vdimir committed:
Add 01944_insert_partition_by to arcadia and fasttest skip list

vdimir committed:
Fix error code in StorageS3.cpp

excitoon commented:
For some reason GitHub does not show your commits separately https://github.com/ClickHouse/ClickHouse/pull/23051/files/82afef5..e91988b

excitoon commented on src/Storages/StorageS3.cpp:
```diff
@@ -322,6 +339,161 @@ class StorageS3BlockOutputStream : public IBlockOutputStream
+                UInt32 invalid_char_byte = static_cast<UInt32>(static_cast<UInt8>(*p));
+                throw DB::Exception(
+                    ErrorCodes::CANNOT_PARSE_TEXT,
+                    "Illegal character '\\x{0:02x}' in partition key. Allowed symbols: [a-zA-Z0-9{1}]", invalid_char_byte, fmt::join(valid_chars, ""));
```
I guess, that this is not very good, alphabetic characters from other languages are disabled now.

excitoon replied:
I would also add links `https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html` and `https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html` to the comments in this case.

excitoon commented on src/Storages/StorageS3.cpp:
```diff
@@ -322,6 +339,161 @@ class StorageS3BlockOutputStream : public IBlockOutputStream
+    static void validateParitionKey(const StringRef & str)
+    {
+        if (str.size == 0)
+            throw DB::Exception(ErrorCodes::CANNOT_PARSE_TEXT, "Illegal empty partition key");
```
Why partition key can not be empty? Hypothetically, it can be a "sorted comma-separated list of binary features".

vdimir replied:
If we allow empty keys such case produce unclear error:


`insert into table function s3('http://localhost:9001/foo/{_partition_id}', 'minioadmin', 'minioadmin', 'CSV', 'id Int32, val String') PARTITION BY val values (1, '');`

```
Received exception from server (version 21.9.1):
Code: 499. DB::Exception: Received from localhost:9000. DB::Exception: Unable to parse ExceptionName: MalformedXML Message: The XML you provided was not well-formed or did not validate against our published schema.. 
```

excitoon replied:
Same in this case:
```
cristal :) insert into table function s3('http://localhost:9001/foo/', 'minioadmin', 'minioadmin', 'CSV', 'id Int32, val String') values (1, '');

INSERT INTO FUNCTION s3('http://localhost:9001/foo/', 'minioadmin', 'minioadmin', 'CSV', 'id Int32, val String') VALUES

Query id: f4c92770-b6fc-45f2-a6a6-5e8af62126b4


1 rows in set. Elapsed: 0.037 sec. 

Received exception from server (version 21.9.1):
Code: 499. DB::Exception: Received from localhost:9000. DB::Exception: Unable to parse ExceptionName: MalformedXML Message: The XML you provided was not well-formed or did not validate against our published schema.. 
```

I think we shall remove any validation in this pull request, BUT implement it in another, in `WriteBufferFromS3`/`ReadBufferFromS3`.

excitoon commented on src/Storages/StorageS3.cpp:
```diff
@@ -322,6 +339,161 @@ class StorageS3BlockOutputStream : public IBlockOutputStream
+        return writers[partition_id];
+    }
+
+    static void validateParitionKey(const StringRef & str)
```
Minor typo — `validatePar`**`t`**`itionKey`.

vdimir commented on src/Storages/StorageS3.cpp:
```diff
@@ -322,6 +339,161 @@ class StorageS3BlockOutputStream : public IBlockOutputStream
+
+    static void validateParitionKey(const StringRef & str)
+    {
+        if (str.size == 0)
```
Also we have to check maximal length

excitoon committed:
Less strict validation.

excitoon committed:
Improved validation of S3 buckets and keys.

excitoon committed:
Update.

excitoon committed:
Build and style fix.

excitoon committed:
Test fix.

excitoon committed:
Test fix.

excitoon commented:
@vdimir Test is fixed but there is something wrong with RabbitMQ in integration tests.

excitoon commented:
@vdimir Can you look through it again, I made some rework after breaking changes.

excitoon committed:
Minor fix.

vdimir commented:
> Test is fixed but there is something wrong with RabbitMQ in integration tests

RabbitMQ tests broken in master

> Can you look through it again? I made some rework after breaking changes

Ok, I'm going to look at the changes

vdimir commented on src/Storages/StorageS3.cpp:
```diff
@@ -327,6 +345,167 @@ class StorageS3Sink : public SinkToStorage
+    {
+        const char * end = str.data + str.size;
+
+        const char * first_symbol = find_first_symbols<
```
But actually simple loop and checking `*pos < 32 && *pos != '{' && *pos != '}' && ...` or something like this will be better, because in current implementation we perform 3 iteration over whole string on "happy path"

excitoon replied:
```
vladimir@cristal:~/find_first$ cat 1.cpp 
int main(int argc, char ** argv)
{
    bool result;
    for (int j = 0; j < 100000000; ++j)
    {
        result = true;
        for (char * i = argv[0]; *i; ++i)
        {
            if (*i < 32 || *i == '{' || *i == '}' || *i == '*' || *i == '?')
            {
                result = false;
                break;
            }
        }
    }
    return result ? 0 : 1;
}

```

```
#include "/home/vladimir/ClickHouse9/base/common/find_symbols.h"

int main(int argc, char ** argv)
{
    bool result;
    size_t len = 0;
    for (; argv[0][len] == 0; ++len);
    char *end = argv[0]+len;
    for (int j = 0; j < 100000000; ++j)
    {
        result = true;

        const char * first_symbol = find_first_symbols<
                '\x00', '\x01', '\x02', '\x03', '\x04', '\x05', '\x06', '\x07',
                '\x08', '\x09', '\x0A', '\x0B', '\x0C', '\x0D', '\x0E', '\x0F'>(argv[0], end);
        if (first_symbol == end)
        {
            first_symbol = find_first_symbols<
                    '\x10', '\x11', '\x12', '\x13', '\x14', '\x15', '\x16', '\x17',
                    '\x18', '\x19', '\x1A', '\x1B', '\x1C', '\x1D', '\x1E', '\x1F'
                >(argv[0], end);
            if (first_symbol == end)
            {
                first_symbol = find_first_symbols<'{', '}', '*', '?'>(argv[0], end);
                if (first_symbol == end)
                {
                    result = false;
                }
            }
        }
    }
    return result ? 0 : 1;
}
```

```
vladimir@cristal:~/find_first$ g++ -std=c++17 -O2 1.cpp -o 1
vladimir@cristal:~/find_first$ g++ -std=c++17 -O2 2.cpp -o 2
vladimir@cristal:~/find_first$ time ./1 abcdefghjkja

real	0m0.378s
user	0m0.367s
sys	0m0.000s
vladimir@cristal:~/find_first$ time ./2 abcdefghjkja

real	0m0.115s
user	0m0.114s
sys	0m0.000s
```

excitoon replied:
Okay

vdimir commented on src/Storages/StorageS3.cpp:
```diff
@@ -327,6 +345,167 @@ class StorageS3Sink : public SinkToStorage
+                context,
+                compression_method,
+                client,
+                replaceWildcards(bucket, partition_id),
+                replaceWildcards(key, partition_id),
```
Should we check that result bucket name and key are not empty? And also maximal length?

excitoon replied:
@vdimir Bucket name is being checked in `S3::URI`, and empty key is allowed in AWS.

vdimir replied:
Do you mean that it's checked on server-side?

```
INSERT INTO FUNCTION s3('http://localhost:9001/foo/{_partition_id}', 'minioadmin', 'minioadmin', 'CSV', 'id Int32, val String') PARTITION BY val VALUES

Received exception from server (version 21.9.1):
Code: 499. DB::Exception: Received from localhost:9000. DB::Exception: Unable to parse ExceptionName: MalformedXML Message: The XML you provided was not well-formed or did not validate against our published schema.. (S3_ERROR)
```


```
INSERT INTO FUNCTION s3('http://localhost:9001/{_partition_id}/key.csv', 'minioadmin', 'minioadmin', 'CSV', 'id Int32, val String') PARTITION BY val VALUES

Received exception from server (version 21.9.1):
Code: 499. DB::Exception: Received from localhost:9000. DB::Exception: Unable to parse ExceptionName: MalformedXML Message: The XML you provided was not well-formed or did not validate against our published schema.. (S3_ERROR)
```

vdimir reviewed (approved):
Ok, only couple of comments

excitoon committed:
Update src/Storages/StorageS3.cpp

Co-authored-by: Vladimir C <vdimir@yandex-team.ru>

excitoon commented on src/Storages/StorageS3.cpp:
```diff
@@ -327,6 +345,167 @@ class StorageS3Sink : public SinkToStorage
+            ));
+        }
+
+        return sinks[partition_id];
```
+

excitoon committed:
Fixed `find`.

excitoon committed:
Style fix.

excitoon committed:
Fixed test.

excitoon commented:
@vdimir Looks like it fixed everything. What are next steps?

excitoon committed:
Fixes.

excitoon committed:
Minor.

vdimir commented on src/Storages/StorageS3.cpp:
```diff
@@ -327,6 +344,156 @@ class StorageS3Sink : public SinkToStorage
+            {
+                /// Need to convert to UInt32 because UInt8 can't be passed to format due to "mixing character types is disallowed".
+                UInt32 invalid_char_byte = static_cast<UInt32>(static_cast<UInt8>(*i));
+                throw DB::Exception(ErrorCodes::CANNOT_PARSE_TEXT,
+                        "Illegal character '\\x{0:02x}' in partition key starting with '{}'.",
+                        invalid_char_byte,
+                        StringRef(str.data, i - str.data));
```
Strange error `fmt::v7::format_error: cannot switch from manual to automatic argument indexing.`, maybe `StringRef` not supported here
```suggestion
                throw DB::Exception(ErrorCodes::CANNOT_PARSE_TEXT,
                        "Illegal character '\\x{0:02x}' in partition key",
                        invalid_char_byte);
```

excitoon replied:
This is not because of `StringRef`.

excitoon replied:
Actually, new code does not work at all.

vdimir committed:
Update src/Storages/StorageS3.cpp

StringRef in fmt::format throws `fmt::v7::format_error: cannot switch from manual to automatic argument indexing`

excitoon committed:
Revert "Update src/Storages/StorageS3.cpp
"

This reverts commit bbbfe1891bfdc868878cfc8d52c254babd4ca181.

excitoon committed:
Better fix.

excitoon committed:
Typo.

excitoon committed:
Added integration test.

vdimir commented:
> @vdimir Looks like it fixed everything. What are next steps?

@excitoon Ok, I'll merge when CI passed. Also look at this comment about checking length https://github.com/ClickHouse/ClickHouse/pull/23051#discussion_r678997963

excitoon committed:
Fix.

excitoon committed:
Validation.

excitoon committed:
Minor fix.

excitoon committed:
Rolled back changes in `Functions`.

excitoon committed:
Rolled back changes in `Functions`.

excitoon committed:
Rolled back changes in `Functions`.

excitoon commented:
@vdimir Fixed it, but there are few failing unrelated tests.

vdimir commented on src/IO/S3Common.cpp:
```diff
@@ -616,55 +617,86 @@ namespace S3
+        /// - https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html
+        /// - https://cloud.ibm.com/apidocs/cos/cos-compatibility#putobject
+
+#if 0
+        /// The following is valid for AWS S3:
+        if (key.length() > 1024)
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Key name is too long (showing first 1024 characters): {}", quoteString(key.substr(0, 1024) + "..."));
+        if (!ValidUTF8Impl::isValidUTF8(reinterpret_cast<const UInt8 *>(key.data()), key.size()))
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Key name must be valid UTF-8 string: {}", quoteString(key));
+#endif
+
+#if 0
+        /// The following is valid for IBM COS:
+        if (key.length() < 1)
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Key name is too short (0 bytes long)");
+#endif
```
Need to be removed?

excitoon replied:
I would leave it there as FIXME/documentation. I guess we may implement more strict checks at some point.

vdimir replied:
But now we wont check that input is UTF8?

excitoon replied:
Yes, because in COS non-UTF-8 is allowed. I guess in separate PR we may implement different checks depending on type of backend

vdimir replied:
So, do we want to leave it without validation? I remember that we decided to validate utf8 + no ascii control chars + no glob chars + length. All is checked except utf8.

excitoon replied:
Yes, because otherwise COS would be broken 🤷‍♂️

excitoon commented:
I made a rebase in order to pass integration tests.

excitoon commented:
@vdimir What are our next steps?

vdimir committed:
Move isValidUTF8Naive from Functions to Common

vdimir committed:
Validate uft8 in partition key from PARTITION BY for s3

vdimir committed:
Do not allow slashes in bucket formatted from PARTITION BY

vdimir committed:
Generate ya.make for isValidUTF8.cpp

vdimir commented:
> @vdimir What are our next steps?

I suppose we should add validation as discussed earlier. I added it and rebased to master, let's wait for CI and merge.

vdimir committed:
Remove unused Error Code from IO/S3Common.cpp

excitoon commented:
@vdimir We can not add validation, we would break `COS` table function by these changes. :man_shrugging:

vdimir committed:
Fix validateKey/Bucket for S3

vdimir commented:
> @vdimir We can not add validation, we would break `COS` table function by these changes. 🤷‍♂️

Validation would be performed only for new feature.

excitoon commented:
@vdimir What is the logic behind that this feature will not support COS fully, though we can easily achieve that?

vdimir committed:
Merge branch 'master' into s3partitionedwrite

vdimir commented:
Commit https://github.com/ClickHouse/ClickHouse/pull/23051/commits/98acccbdcc39ec95a22626b7b256489474338dc2 is only merged master to resolve conflicts, no changes, build in Fast test passed, will merge. Full CI passed on commit https://github.com/ClickHouse/ClickHouse/pull/23051/commits/6fe63a80bc6addbb063074f665f71f2a7842612c

sevirov commented:
Internal documentation ticket: DOCSUP-13742

alexey-milovidov commented:
@excitoon @vdimir an issue has been found: #34348

alexey-milovidov commented:
@excitoon @vdimir a bug has been found: #38199
